### ISSUE TITLE
Snakemake shouldn't invalidate future steps in case some files were changed (devsets, marian binaries, etc..)

### ISSUE THREAD

**Issue by XapaJIaMnu** [mozilla/firefox-translations-training#52]
Currently, if Snakemake detects a change in a file that future steps are dependent on, it will invalidate all previous steps. However, it is not necessarely to do so in all cases. Two examples:
1) marian binary is rebuilt due to system upgrade. The previous trained models and vocabulary files are still valid and shouldn't need invalidation
2) Devset/test set changed. The models built are still valid and shouldn't need to be rebuilt from scratch, just because a different devset was given. If the user feels that the models should be rebuilt from scratch, they can just delete the relevant files by themselves.

In the case of 2), it is also possible that we would want to resume training of existing models, rather than start training them from scratch. This behaviour would also be preferred to the current one.

**Comment by eu9ene** [comment]
This is the default snakemake behaviour and I'm not aware of how to change it atm. I think it makes sense to preserve it because it makes a regular execution safer. If you, say, retrained vocabulary it will be a bug to use it with models trained with another vocabulary, so they should be invalidated.

The pipeline is designed to work end to end to support the ability to train a massive number of languages in a safe way. 

I understand that there are use cases when we want to tweak smth manually, especially on debugging and research exploration. IMO these manual interventions should require other manual actions to make the workflow behave as you want in a specific case.

For the specified use cases:
1. We can add an option to use pre-compiled marian (like backward model and vocabulary) and remove dependency to this rule. I think if a file is already there, snakemake won't reschedule the step and invalidate future steps. However, it means that we'll need to recompile it separately.
2. Devset (validation set) is a part of retraining and its purpose is only to validate models while training, so I don't think there is a use case when you want to change a devset without retraining models. Test sets are used for evaluation only and can be safely updated on any step - all evaluation steps will just rerun and they are fast. Changing any of this in the middle of training is not desirable and not supported. 

In general, I think we need to provide some tools that are capable of instructing snakemake to skip the specified steps. They can be useful for debugging, bugfixing in the middle of the training and any other weird edge cases.

**Comment by XapaJIaMnu** [comment]
I opened this issue because i run into the following issues:

1) There are no official test sets for a lot of low resource language. I manually constructed a test set, but it turned out to have issues. I updated the test set and snakemake wanted to restart the training. I couldn't find a way around that.
2) A lot of super computing providers limit the length of GPU jobs and you would inevitably need to restart the training a few times.
3) Further to point 2, sometimes you may need to recompile marian due to change in architecture.

I agree that obv changes to the vocabulary should invalidate further steps, but changing to the spm binary, shouldn't invalidate the vocabulary. There's seldom API breackages when changing the versions of toolkits, and they are designed to check for incompatible APIs.

**Comment by eu9ene** [comment]
I found snakemake option that will solve recompilation issues: https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#ignoring-timestamps

We can mark compiled input as ancient

**Comment by eu9ene** [comment]
`ancient` was added for all binaries, there will be no cascade rerunning after recompilation. I think the main concern of this issue is resolved and we can close it. 

If a bug was found in the middle of training and a user wants to fix something like dev dataset but skip cascade rerunning, keyword `ancient` should be used ad hoc for this specific situation. 

Default Snakemake behaviour should remain the same to prevent unexpected bugs that are hard to find like the one with accidentally retrained vocabulary.
